common/ipmi/storage_handler.c

```diff
@@ -0,0 +1,225 @@
+		break;
+	case FRU_FAIL_TO_ACCESS:
+		msg->completion_code = CC_FRU_DEV_BUSY;
+		break;
+	default:
+		msg->completion_code = CC_UNSPECIFIED_ERROR;
+		break;
+	}
+
+	return;
+}
+
+__weak void STORAGE_WRITE_FRUID_DATA(ipmi_msg *msg)
+{
+	uint8_t status;
+	EEPROM_ENTRY fru_entry;
+
+	if (msg->data_len < 4) {
+		msg->completion_code = CC_INVALID_LENGTH;
+		return;
+	}
+
+	fru_entry.config.dev_id = msg->data[0];
+	fru_entry.offset = (msg->data[2] << 8) | msg->data[1];
+	fru_entry.data_len = msg->data_len - 3; // skip id and offset
+	if (fru_entry.data_len > 32) { // According to IPMI, messages are limited to 32 bytes
+		msg->completion_code = CC_LENGTH_EXCEEDED;
+		return;
+	}
+	memcpy(&fru_entry.data[0], &msg->data[3], fru_entry.data_len);
+
+	msg->data[0] = msg->data_len - 3;
+	msg->data_len = 1;
+	status = FRU_write(&fru_entry);
+
+	switch (status) {
+	case FRU_WRITE_SUCCESS:
+		msg->completion_code = CC_SUCCESS;
+		break;
+	case FRU_INVALID_ID:
+		msg->completion_code = CC_INVALID_PARAM;
+		break;
+	case FRU_OUT_OF_RANGE:
+		msg->completion_code = CC_PARAM_OUT_OF_RANGE;
+		break;
+	case FRU_FAIL_TO_ACCESS:
+		msg->completion_code = CC_FRU_DEV_BUSY;
+		break;
+	default:
+		msg->completion_code = CC_UNSPECIFIED_ERROR;
+		break;
+	}
+
+	return;
+}
+
+__weak void STORAGE_RSV_SDR(ipmi_msg *msg)
+{
+	uint16_t RSV_ID;
+
+	if (msg->data_len != 0) {
+		msg->completion_code = CC_INVALID_LENGTH;
+		return;
+	}
+
+	RSV_ID = SDR_get_RSV_ID();
+	msg->data[0] = RSV_ID & 0xFF;
+	msg->data[1] = (RSV_ID >> 8) & 0xFF;
+	msg->data_len = 2;
+	msg->completion_code = CC_SUCCESS;
+
+	return;
+}
+
+__weak void STORAGE_GET_SDR(ipmi_msg *msg)
+{
+	uint16_t next_record_ID;
+	uint16_t rsv_ID, record_ID;
+	uint8_t offset, req_len;
+	uint8_t *table_ptr;
+
+	rsv_ID = (msg->data[1] << 8) | msg->data[0];
+	record_ID = (msg->data[3] << 8) | msg->data[2];
+	offset = msg->data[4];
+	req_len = msg->data[5];
+
+	if (msg->data_len != 6) {
+		msg->completion_code = CC_INVALID_LENGTH;
+		return;
+	}
+
+	if (!SDR_RSV_ID_check(rsv_ID)) {
+		msg->completion_code = CC_INVALID_RESERVATION;
+		return;
+	}
+
+	if (!SDR_check_record_ID(record_ID)) {
+		msg->completion_code = CC_INVALID_DATA_FIELD;
+		return;
+	}
+
+	// request length + next record ID should not over IPMB data limit
+	if ((req_len + 2) > IPMI_DATA_MAX_LENGTH) {
+		msg->completion_code = CC_LENGTH_EXCEEDED;
+		return;
+	}
+
+	if ((offset + req_len) > IPMI_SDR_HEADER_LEN + full_sensor_table[record_ID].record_len) {
+		msg->completion_code = CC_PARAM_OUT_OF_RANGE;
+		return;
+	}
+
+	next_record_ID = SDR_get_record_ID(record_ID);
+	msg->data[0] = next_record_ID & 0xFF;
+	msg->data[1] = (next_record_ID >> 8) & 0xFF;
+
+	table_ptr = (uint8_t *)&full_sensor_table[record_ID];
+	memcpy(&msg->data[2], (table_ptr + offset), req_len);
+
+	msg->data_len = req_len + 2; // return next record ID + sdr data
+	msg->completion_code = CC_SUCCESS;
+
+	return;
+}
+
+void IPMI_Storage_handler(ipmi_msg *msg)
+{
+	switch (msg->cmd) {
```

Comment:
msg == NULL check?

Reply:
We have fixed it and done the forced update.
Please review it. Thanks.